source/extensions/health_checkers/mysql/mysql.cc

```diff
@@ -0,0 +1,323 @@
+  const auto lower_capabilities =
+      read_little_endian_int<lower_capabilities_size>(packet + lower_capabilities_offset);
+  if (!(lower_capabilities & MYSQL_CAPABILITIES_CLIENT_PROTOCOL_41)) {
+    ENVOY_CONN_LOG(
+        trace, "mysql healthcheck failed [greeting]: server does not support 4.1 client protocol ",
+        *client_);
+    return false;
+  }
+
+  return true;
+}
+
+void MySQLHealthChecker::MySQLActiveHealthCheckSession::write_mysql_login_request_packet(
+    Buffer::Instance& buffer) {
+  // write login request packet:
+  // https://dev.mysql.com/doc/internals/en/connection-phase-packets.html
+  //  #packet-Protocol::HandshakeResponse320
+
+  constexpr const uint32_t MYSQL_CAPABILITIES_CLIENT_CAN_DO_41_AUTH = 0x8000;
+  constexpr const size_t client_capabilities_size = 2;
+  constexpr const size_t max_packets_size = 3;
+  const size_t user_size = parent_.user_.size() + 1;
+  constexpr const size_t auth_response_size = 1;
+  const uint8_t nul = '\0';
+  const size_t auth_response = '\0';
+
+  // the extra +1 accounts for the NUL terminator in the user name
+  const size_t packet_size =
+      client_capabilities_size + max_packets_size + user_size + auth_response_size;
+  write_little_endian_int<mysql_payload_length_size>(buffer, packet_size);
+
+  constexpr const auto sequence_id = 1;
+  write_little_endian_int<mysql_sequence_id_size>(buffer, sequence_id);
+
+  constexpr const auto client_capabilities = MYSQL_CAPABILITIES_CLIENT_CAN_DO_41_AUTH;
+  write_little_endian_int<client_capabilities_size>(buffer, client_capabilities);
+
+  constexpr const auto max_packets = 65536;
+  write_little_endian_int<max_packets_size>(buffer, max_packets);
+
+  buffer.add(parent_.user_.data(), parent_.user_.size());
+  buffer.add(&nul, 1);
+
+  buffer.add(&auth_response, auth_response_size);
+}
+
+void MySQLHealthChecker::MySQLActiveHealthCheckSession::write_mysql_quit_packet(
+    Buffer::Instance& buffer) {
+  // write quit packet:
+  // https://dev.mysql.com/doc/internals/en/com-quit.html
+  //  #packet-COM_QUIT
+
+  constexpr const auto MYSQL_QUIT_COMMAND = 1;
+  constexpr const size_t quit_command_size = 1;
+
+  const size_t packet_size = quit_command_size;
+  write_little_endian_int<mysql_payload_length_size>(buffer, packet_size);
+
+  constexpr const auto sequence_id = 0;
+  write_little_endian_int<mysql_sequence_id_size>(buffer, sequence_id);
+
+  write_little_endian_int<quit_command_size>(buffer, MYSQL_QUIT_COMMAND);
+}
+
+bool MySQLHealthChecker::MySQLActiveHealthCheckSession::parse_mysql_ok_packet(
+    Buffer::Instance& data) {
+  // parse response packet
+  // https://dev.mysql.com/doc/internals/en/packet-OK_Packet.html
+  constexpr const auto MYSQL_OK_RESPONSE = 0;
+  constexpr const size_t ok_response_size = 1;
+  constexpr const size_t ok_packet_size_threshold = 7;
+
+  const size_t minimum_needed_packet_size = mysql_packet_header_size + ok_response_size;
+  if (data.length() < minimum_needed_packet_size) {
```

Comment:
Technically, I think you need to handle partial data here (and elsewhere in similar parsing functions)? There is no guarantee that all the data will come in 1 shot. If that is something that you don't think we need to handle, we should clearly document that and possibly add a TODO here, though I would rather we just handle it. Thoughts?

Reply:
I haven't seen the need for it just yet, so I believe leaving it as a TODO might suffice. As for how partial data is handled elsewhere in Envoy, does it accumulate partial data in a local buffer and parses in one shot or is there support for a parsing state machine? Implementing such a machine by hand sounds counter-productive and error prone.